Origin: Commit - 4978f3dda1c070f5680923c2967fdae135678ee2
Author: Campbell Jones <git@serebit.com>
Last-Update: 2022-05-14
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/budgie-desktop/+bug/1990642
Description: [PATCH] Rewrite initial applet layout to prevent position desync

---
 src/panel/panel.vala | 51 +++++++++++++++++++++++++++++++++++---------
 1 file changed, 41 insertions(+), 10 deletions(-)

diff --git a/src/panel/panel.vala b/src/panel/panel.vala
index 20fbc862..64cb35a9 100644
--- a/src/panel/panel.vala
+++ b/src/panel/panel.vala
@@ -514,12 +514,20 @@ namespace Budgie {
 				return;
 			}
 
+			CompareFunc<Budgie.AppletInfo?> infocmp = (a, b) => {
+				return (int) (a.position > b.position) - (int) (a.position < b.position);
+			};
+
 			/* Two loops so we can track when we've fully loaded the panel */
 			lock (expected_uuids) {
 				for (int i = 0; i < applets.length; i++) {
 					this.expected_uuids.append(applets[i]);
 				}
 
+				var start_applets = new List<Budgie.AppletInfo?>();
+				var center_applets = new List<Budgie.AppletInfo?>();
+				var end_applets = new List<Budgie.AppletInfo?>();
+
 				for (int i = 0; i < applets.length; i++) {
 					string? name = null;
 					Budgie.AppletInfo? info = this.manager.load_applet_instance(applets[i], out name);
@@ -533,15 +541,38 @@ namespace Budgie {
 								expected_uuids.remove_link(g);
 							}
 							message("Unable to load invalid applet: %s", applets[i]);
+							applet_removed(applets[i]);
+							continue;
+						}
+
+						info = this.add_pending(applets[i], name);
+						if (info == null) {
 							continue;
 						}
-						this.add_pending(applets[i], name);
-						manager.modprobe(name);
+					}
+
+					/* um add this bro to the panel :o */
+					if (info.alignment == "start") {
+						start_applets.insert_sorted(info, infocmp);
+					} else if (info.alignment == "center") {
+						center_applets.insert_sorted(info, infocmp);
 					} else {
-						/* um add this bro to the panel :o */
-						this.add_applet(info);
+						end_applets.insert_sorted(info, infocmp);
 					}
 				}
+
+				for (int i = 0; i < start_applets.length(); i++) {
+					start_applets.nth_data(i).position = i;
+					add_applet(start_applets.nth_data(i));
+				}
+				for (int i = 0; i < center_applets.length(); i++) {
+					center_applets.nth_data(i).position = i;
+					add_applet(center_applets.nth_data(i));
+				}
+				for (int i = 0; i < end_applets.length(); i++) {
+					end_applets.nth_data(i).position = i;
+					add_applet(end_applets.nth_data(i));
+				}
 			}
 		}
 
@@ -899,13 +930,13 @@ namespace Budgie {
 			return;
 		}
 
-		void add_pending(string uuid, string plugin_name) {
+		Budgie.AppletInfo? add_pending(string uuid, string plugin_name) {
 			string? rname = null;
 			unowned HashTable<string,string>? table = null;
 
 			if (!this.manager.is_extension_valid(plugin_name)) {
 				warning("Not adding invalid plugin: %s %s", plugin_name, uuid);
-				return;
+				return null;
 			}
 
 			if (!this.manager.is_extension_loaded(plugin_name)) {
@@ -915,23 +946,23 @@ namespace Budgie {
 					if (!table.contains(uuid)) {
 						table.insert(uuid, uuid);
 					}
-					return;
+					return null;
 				}
 				/* Looks insane but avoids copies */
 				pending.insert(plugin_name, new HashTable<string,string>(str_hash, str_equal));
 				table = pending.lookup(plugin_name);
 				table.insert(uuid, uuid);
 				this.manager.modprobe(plugin_name);
-				return;
+				return null;
 			}
 
 			/* Already exists */
 			Budgie.AppletInfo? info = this.manager.load_applet_instance(uuid, out rname);
 			if (info == null) {
 				critical("Failed to load applet when we know it exists");
-				return;
+				return null;
 			}
-			this.add_applet(info);
+			return info;
 		}
 
 		public override void map() {
-- 
2.34.1

